docs/Upgrade.md

```diff
@@ -482,6 +482,109 @@ import { FieldProps, useRecordContext } from 'react-admin';
+);
+
++const router = createBrowserRouter([{ path: '*', element: <App /> }]);
+
+const container = document.getElementById('root');
+const root = createRoot(container);
+
+root.render(
+    <React.StrictMode>
+-        <App />
++        <RouterProvider router={router} />
+    </React.StrictMode>
+);
+```
+
+### Minor Changes
+
+Due to the new implementation using `useBlocker`, you may also notice the following minor changes:
+
+- `useWarnWhenUnsavedChanges` no longer waits until the submission is complete before redirecting. Instead, it blocks navigation and opens the confirmation dialog if the form is currently submitting. While the dialog is open, submission continues in the background.
+- Implementing blocking when closing the browser tab now requires listening to the `'beforeunload'` event. [Modern browsers no longer allow to customize the message](https://stackoverflow.com/questions/38879742/is-it-possible-to-display-a-custom-message-in-the-beforeunload-popup) displayed in the confirmation dialog, so users will see the default message from the browser in that case.
+
+## `<Admin history>` Prop Was Removed
+
+The `<Admin history>` prop was deprecated since version 4. It is no longer supported.
+
+The most common use-case for this prop was inside unit tests (and stories), to pass a `MemoryRouter` and control the `initialEntries`.
+
+To that purpose, `react-admin` now exports a `TestMemoryHistory` component that you can use in your tests:
+
+```diff
+import { render, screen } from '@testing-library/react';
+-import { createMemoryHistory } from 'history';
+-import { CoreAdminContext } from 'react-admin';
++import { CoreAdminContext, TestMemoryRouter } from 'react-admin';
+import * as React from 'react';
+
+describe('my test suite', () => {
+    it('my test', async () => {
+-       const history = createMemoryHistory({ initialEntries: ['/'] });
+        render(
++           <TestMemoryRouter initialEntries={['/']}>
+              <CoreAdminContext history={history}>
+                <div>My Component</div>
+              </CoreAdminContext>
++           </TestMemoryRouter>
+        );
+        await screen.findByText('My Component');
+    });
+});
+```
+
+### Codemod
+
+To help you migrate your tests, we've created a codemod that will replace the `<Admin history>` prop with the `<TestMemoryRouter>` component.
+
+> **DISCLAIMER**
+>
+> This codemod was used to migrate the react-admin test suite, but it was never designed to cover all cases, and was not tested against other code bases. You can try using it as basis to see if it helps migrating your code base, but please review the generated changes thoroughly!
+>
+> Applying the codemod might break your code formatting, so please don't forget to run `prettier` and/or `eslint` after you've applied the codemod!
+
+For `.js` or `.jsx` files:
+
+```sh
+npx jscodeshift ./path/to/src/ \
+    --extensions=js,jsx \
+    --transform=./node_modules/ra-core/codemods/replace-Admin-history.ts
```

Comment:
There are 2 codemods, you only document one. By the way, shouldn't we present only one codemode doing everything?

Reply:
Indeed.

- the `replace-Admin-history.ts` codemod helps replacing the `history` prop by a `<TestMemoryRouter>`, which is what I document in the Upgrade guide anyway
- the `replace-MemoryRouter.ts` codemod is required to replace instances of `<MemoryRouter>`, which is **not** a data router, by `<TestMemoryRouter>`, which is a data router. Technically this change is only required for tests (and stories) including a `<Form>`.

I figured I could not include this last codemod in the Upgrade guide without getting too much into technical details about why this is needed. So for the sake of simplicity I chose to omit it. But I'm open to discussion on the matter.

And, to me, both codemods serve a different purpose, so I thought it would be clearer (and simpler to maintain and to test) to keep them separate. Again, I'm open to discussion here.